fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexerThrottle.java

```diff
@@ -0,0 +1,266 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.foundationdb.record.provider.foundationdb;
+
+import com.apple.foundationdb.FDBError;
+import com.apple.foundationdb.FDBException;
+import com.apple.foundationdb.async.AsyncUtil;
+import com.apple.foundationdb.async.MoreAsyncUtil;
+import com.apple.foundationdb.record.IndexState;
+import com.apple.foundationdb.record.RecordCoreStorageException;
+import com.apple.foundationdb.record.logging.KeyValueLogMessage;
+import com.apple.foundationdb.record.logging.LogMessageKeys;
+import com.apple.foundationdb.record.metadata.Index;
+import com.apple.foundationdb.util.LoggableException;
+import com.google.common.annotations.VisibleForTesting;
+import org.apache.commons.lang3.tuple.Pair;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.annotation.Nonnull;
+import javax.annotation.Nullable;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.BiConsumer;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+/**
+ * This class provides build / commit /retry with throttling to the OnlineIndexer. In the future,
+ * this class can be generalized to serve other FDB modules.
+ *
+ */
+public class OnlineIndexerThrottle {
+
+    @Nonnull private static final Logger LOGGER = LoggerFactory.getLogger(OnlineIndexerThrottle.class);
+    @Nonnull private final OnlineIndexerCommon common;
+    @Nonnull private final AtomicLong totalRecordsScanned;
+
+    private int limit;
+
+    // These error codes represent a list of errors that can occur if there is too much work to be done
+    // in a single transaction.
+    private static final Set<Integer> lessenWorkCodes = new HashSet<>(Arrays.asList(
+            FDBError.TIMED_OUT.code(),
+            FDBError.TRANSACTION_TOO_OLD.code(),
+            FDBError.NOT_COMMITTED.code(),
+            FDBError.TRANSACTION_TIMED_OUT.code(),
+            FDBError.COMMIT_READ_INCOMPLETE.code(),
+            FDBError.TRANSACTION_TOO_LARGE.code()));
+
+    /**
+     * The number of successful transactions in a row as called by {@link #runAsync(Function, BiFunction, BiConsumer, List)}.
+     */
+    private int successCount = 0;
+
+    OnlineIndexerThrottle(@Nonnull OnlineIndexerCommon common) {
+
+        this.common = common;
+        this.totalRecordsScanned = common.getTotalRecordsScanned();
```

Comment:
This feels somewhat clunky. Because it's keeping a copy of the reference (essentially), it means that updates to the throttler's `totalRecordsScanned` also update the `OnlineIndexerCommon`'s copy. I think that's intentional, but buy copying the value to its own instance member, it can make the side effects less obvious. I think it would be better to remove the member and just update the value through the `Common` object.

Reply:
++